pkg/client/linstor.go

```diff
@@ -372,6 +378,101 @@ func (s *Linstor) Create(ctx context.Context, vol *volume.Info, params *volume.P
 	return nil
 }
 
+func (s *Linstor) Clone(ctx context.Context, vol, src *volume.Info, params *volume.Parameters, topologies *csi.TopologyRequirement) error {
+	logger := s.log.WithFields(logrus.Fields{
+		"source": fmt.Sprintf("%+v", src),
+		"volume": fmt.Sprintf("%+v", vol),
+	})
+
+	logger.Debug("reconcile resource group from storage class")
+
+	rGroup, err := s.reconcileResourceGroup(ctx, params)
+	if err != nil {
+		logger.Debugf("reconcile resource group failed: %v", err)
+		return err
+	}
+
+	logger.Debugf("check for clone status")
+
+	status, err := s.client.ResourceDefinitions.CloneStatus(ctx, src.ID, vol.ID)
+	if err != nil && !errors.Is(err, lapi.NotFoundError) {
+		logger.Debugf("clone status failed: %v", err)
+		return err
+	}
+
+	if errors.Is(err, lapi.NotFoundError) {
+		logger.Debugf("create new cloned volume")
+
+		_, err := s.client.ResourceDefinitions.Clone(ctx, src.ID, lapi.ResourceDefinitionCloneRequest{
+			Name:          vol.ID,
+			LayerList:     params.LayerList,
+			ResourceGroup: rGroup.Name,
+		})
+		if err != nil {
+			logger.Debugf("clone failed: %v", err)
+			return err
+		}
+
+		status, err = s.client.ResourceDefinitions.CloneStatus(ctx, src.ID, vol.ID)
+		if err != nil {
+			logger.Debugf("clone status failed: %v", err)
+			return err
+		}
+	}
+
+	// This is not perfect: if we fail to delete the ProvisioningCompletedBy property, the volume creation will be
+	// retried. If LINSTOR has already finished cloning the volume, we could think that the volume is already "done"
+	// with provisioning.
+	//
+	// Mitigating factors: it is extremely unlikely that we fail exactly at the step below and that cloning
+	// completes faster than the CSI retry.
+	logger.Debugf("Reset ProvisioningCompletedBy for new resource definition")
+
+	err = s.client.ResourceDefinitions.Modify(ctx, vol.ID, lapi.GenericPropsModify{DeleteProps: []string{linstor.PropertyProvisioningCompletedBy}})
+	if err != nil {
+		logger.Debugf("failed resetting cloned resource definition property: %v", err)
+		return err
+	}
+
+	for status.Status != clonestatus.Complete {
+		logger.Debug("clone in progress, wait 5 seconds")
+
+		time.Sleep(5 * time.Second)
+
+		status, err = s.client.ResourceDefinitions.CloneStatus(ctx, src.ID, vol.ID)
+		if err != nil {
+			logger.Debugf("clone status failed: %v", err)
+			return err
+		}
+	}
+
+	logger.Debug("reconcile volume definition for volume")
+
+	_, err = s.reconcileVolumeDefinition(ctx, vol)
+	if err != nil {
+		logger.Debugf("reconcile volume definition failed: %v", err)
+		return err
+	}
+
+	logger.Debug("reconcile volume placement")
+
+	err = s.reconcileResourcePlacement(ctx, vol, params, topologies)
+	if err != nil {
```

Comment:
personally for these cases I prefer `if err := call(); err != nil {` but that is probably just a matter of taste, so feel free to ignore